### ISSUE TITLE
GeneratorDrivenPropertyCheck failing on Scala-js 2.10 in 3.0.0-M7

### ISSUE THREAD

**Issue by InTheNow** [scalatest/scalatest#661]
We are successfully using scalatest in [Cats](https://github.com/non/cats), so firstly: Thanks!!

There is just one issue that pops up occasionally and I wonder if you could investigate. It only happens on Travis, on scala.js, in 2.10 - [for example, this build](https://travis-ci.org/non/cats/jobs/75846115#L4075) from https://github.com/non/cats/pull/453. As travis output can mangled:

```sbt
org.scalatest.exceptions.GeneratorDrivenPropertyCheckFailedException: Gave up after 0 successful property evaluations. 11 evaluations were discarded.
	at scala.scalajs.runtime.StackTrace$.createException(/home/travis/build/non/cats/tests/js/target/scala-2.10/cats-tests-test-fastopt.js:26286)
	at java.lang.Throwable.fillInStackTrace(/home/travis/build/non/cats/tests/js/target/scala-2.10/cats-tests-test-fastopt.js:35173)
	at java.lang.Throwable.<init>(/home/travis/build/non/cats/tests/js/target/scala-2.10/cats-tests-test-fastopt.js:35193)
	at java.lang.RuntimeException.<init>(/home/travis/build/non/cats/tests/js/target/scala-2.10/cats-tests-test-fastopt.js:49611)
	at org.scalatest.exceptions.StackDepthException.<init>(/home/travis/build/non/cats/tests/js/target/scala-2.10/cats-tests-test-fastopt.js:72130)
	at org.scalatest.exceptions.TestFailedException.<init>(/home/travis/build/non/cats/tests/js/target/scala-2.10/cats-tests-test-fastopt.js:93380)
	at org.scalatest.exceptions.PropertyCheckFailedException.<init>(/home/travis/build/non/cats/tests/js/target/scala-2.10/cats-tests-test-fastopt.js:94027)
	at org.scalatest.exceptions.GeneratorDrivenPropertyCheckFailedException.<init>(/home/travis/build/non/cats/tests/js/target/scala-2.10/cats-tests-test-fastopt.js:94589)
	at org.scalatest.prop.Checkers$.doCheck(/home/travis/build/non/cats/tests/js/target/scala-2.10/cats-tests-test-fastopt.js:40984)
	at org.scalatest.prop.Checkers$class.check(/home/travis/build/non/cats/tests/js/target/scala-2.10/cats-tests-test-fastopt.js:15305)
[error] Could not run test cats.tests.IorTests: org.scalatest.exceptions.GeneratorDrivenPropertyCheckFailedException: Gave up after 0 successful property evaluations. 11 evaluations 
```

[This is the failing test](https://github.com/non/cats/blob/master/tests/shared/src/test/scala/cats/tests/IorTests.scala).

Also note that we override test parameters for [js](https://github.com/non/cats/blob/master/tests/js/src/test/scala/cats/tests/Platform.scala) and [jvm](https://github.com/non/cats/blob/master/tests/jvm/src/test/scala/cats/tests/Platform.scala).

Now, I concede that there are many of reasons why the test might fail, but I am particularly concerned that it is exhausted on the 11th evaluation, despite setting the discard ratio very high. 

Is it possible that this is all down to [a hard coded value](https://github.com/scalatest/scalatest/blob/3.0.x/scalatest/src/main/scala/org/scalatest/prop/GenDrivenPropertyChecks.scala#L28)? And if not, any other ideas, please?

**Comment by bvenners** [comment]
No, that comment is about a class that's private and not being used. Where are you setting the discard ratio? (I.e., what are you setting it to?)

**Comment by InTheNow** [comment]
it is set in https://github.com/non/cats/blob/master/tests/js/src/test/scala/cats/tests/Platform.scala 

See "Also note that we override test parameters for js and jvm" above, the ratio is set to 50

**Comment by bvenners** [comment]
Ok, the name of this file made me laugh: https://github.com/non/cats/blob/master/project/Boilerplate.scala

Looking for your build to see what version of ScalaTest you're using...

**Comment by InTheNow** [comment]
Here is a print output:
```
class IorTests extends CatsSuite {
  println(generatorDrivenConfig)
```

gives 

```
testsJS/testOnly cats.tests.IorTests
PropertyCheckConfiguration(PosInt(10),PosZDouble(50),PosZInt(0),PosZInt(100),PosInt(1))
[info] IorTests:
[info] - Ior[String, Int].monad.applicative homomorphism
```

**Comment by bvenners** [comment]
One thing I think we need to do is let 50.0 be the default for ratio, since we reduced the minSuccessful to 10. Chee Seng actually submitted that, but I took it out. Both Erik Osheim and Mike Pilquist said they hit too many discarded problem on code that used to work. But on the other hand I wonder if something else is going on. Perhaps we've messed this up:

https://github.com/scalatest/scalatest/blob/3.0.x/scalatest/src/main/scala/org/scalatest/prop/Configuration.scala#L438

I unfortunately need to get ready for a visit to the tax preparer this today. I'll try and look at it later, but will also point Brian Roberts to this, as he did this particular refactor for us.

**Comment by flicken** [comment]
At first glance, the code looks like it's applying the conversion correctly.  It looks like `Parameters.maxDiscardRatio` should end up being `config.maxDiscardedFactor.value`, or 50.0 in this case.  Next step would be to ensure that `Parameter` values are as we expect.

Unless someone beats me to it, I'll have time later this week or early next to look into further.

**Comment by InTheNow** [comment]
@flicken great - thanks for the feedback

**Comment by InTheNow** [comment]
Hi guys! Just doing some paperwork -  Any developments on this and any ideas on times for a new Milestone release, please?

**Comment by bvenners** [comment]
Hi Alistair,

Sorry we haven't had any time to look at this yet. The contract we're all working on ends Monday, though, so if we don't get to it before that, we'll do it first thing after.

Bill

**Comment by cheeseng** [comment]
@InTheNow @bvenners @flicken I think I figure out why it failed at after 11 discarded, because as you mentioned in js, you set the minSuccessful to 10: 

https://github.com/non/cats/blob/master/tests/js/src/test/scala/cats/tests/Platform.scala#L9

and here's the formula scalacheck is using to decide whether to give up (n = success, d = discarded): 

n+d > minSuccessfulTests && 1+workers*maxDiscardRatio*n < d  (see https://github.com/rickynils/scalacheck/blob/master/src/main/scala/org/scalacheck/Test.scala#L292)

so in your failing case, n = 0, d = 11, minSuccessfulTests = 10, maxDiscardRatio = 50 which is: 

n+d > minSuccessfulTests && 1+workers*maxDiscardRatio*n < d
11 > 10 && 1 + 50 * 0 < 11

Indeed the second part of the condition is already true when discard of the first 1-10, it is the 11th discard that turns the first part to true and consequently signal scalacheck to give up.

Hope this helps.

**Comment by bvenners** [comment]
I submitted an issue to ScalaCheck. This may be be addressed there: https://github.com/rickynils/scalacheck/issues/192

**Comment by bvenners** [comment]
@inthenow can you confirm whether the change Rickard did in ScalaCheck to address this has solved your problem?

**Comment by bvenners** [comment]
@inthenow can you confirm whether the change Rickard did in ScalaCheck to address this has solved your problem? I would like to close this if so...

**Comment by InTheNow** [comment]
@bvenners 

Just to let you know I've not given up on this.... but with all the dependencies in the Typelevel projects not moving forward, this is hard to test without an uber local build of the TL universe.

But as you are looking to release a final 3.0.0 in the near-ish future, I shall do such a build.

@adelbertc   @ceedubs  @non + anyone else. Something to look out for when we got to scalacheck 1.13 is that the current parameter settings we have will need tweaking else we are likely to see builds hanging again.
